kong/router/transform.lua

```diff
@@ -696,34 +675,37 @@ end
 local uuid_generator = assert(uuid.factory_v5('7f145bf9-0dce-4f91-98eb-debbce4b9f6b'))
 
 
-local function sort_by_regex_or_length(path)
-  return is_regex_magic(path) or #path
-end
+-- Turns route.paths array, e.g. { "~/regex.*$", "/long-path", "/one", "two, "/three", "~/.*" } to
+-- a regex/length grouped array: { { "~/regex.*$", "~/.*" }, { "/long-path" }, { "/one", "/two }, { "/three" } }
+local path_groups = {} -- we reuse this to avoid runtime table creation (not thread safe - aka do not yield with it)
+local path_groups_map = {} -- we reuse this to avoid runtime table/garbage creation (not thread safe - aka do not yield with it)
+local function group_by_regex_or_length(paths)
+  tb_clear(path_groups)
+  tb_clear(path_groups_map)
+  local path_groups_count = 0
 
+  for _, path in ipairs(paths) do
+    local k = is_regex_magic(path) and 0 or #path
+    if path_groups_map[k] then
+      tb_insert(path_groups[path_groups_map[k]], path)
 
--- group array-like table t by the function f, returning a table mapping from
--- the result of invoking f on one of the elements to the actual elements.
-local function group_by(t, f)
-  local result = {}
-  for _, value in ipairs(t) do
-    local key = f(value)
-    if result[key] then
-      tb_insert(result[key], value)
     else
-      result[key] = { value }
+      path_groups_count = path_groups_count + 1
+      path_groups_map[k] = path_groups_count
+      path_groups[path_groups_count] = { path }
     end
   end
-  return result
+
+  return path_groups_count, path_groups
```

Comment:
We have the different names where it be called (grouped_paths_count, grouped_paths), could we rename them to be more understandable?

Reply:
@chronolaw then the local variables would have same name as module level variables, which is just confusing (aka reusing module level variable name as local variable name).